adaptive/tests/test_pickling.py

```diff
@@ -0,0 +1,173 @@
+import operator
+import pickle
+import random
+
+import flaky
+import pytest
+
+from adaptive.learner import (
+    AverageLearner,
+    BalancingLearner,
+    DataSaver,
+    IntegratorLearner,
+    Learner1D,
+    Learner2D,
+    LearnerND,
+    SequenceLearner,
+)
+from adaptive.runner import simple
+
+try:
+    import cloudpickle
+
+    with_cloudpickle = True
+except ModuleNotFoundError:
+    with_cloudpickle = False
+
+try:
+    import dill
+
+    with_dill = True
+except ModuleNotFoundError:
+    with_dill = False
+
+
+def goal_1(learner):
+    return learner.npoints == 10
+
+
+def goal_2(learner):
+    return learner.npoints == 20
+
+
+learners_pairs = [
+    (Learner1D, dict(bounds=(-1, 1))),
+    (Learner2D, dict(bounds=[(-1, 1), (-1, 1)])),
+    (LearnerND, dict(bounds=[(-1, 1), (-1, 1), (-1, 1)])),
+    (SequenceLearner, dict(sequence=list(range(100)))),
+    (IntegratorLearner, dict(bounds=(0, 1), tol=1e-3)),
+    (AverageLearner, dict(atol=0.1)),
+]
+
+serializers = [pickle]
+if with_cloudpickle:
+    serializers.append(cloudpickle)
+if with_dill:
+    serializers.append(dill)
+
+learners = [
+    (learner_type, learner_kwargs, serializer)
+    for serializer in serializers
+    for learner_type, learner_kwargs in learners_pairs
+]
+
+
+def f_for_pickle(x):
+    return 1
+
+
+def f_for_pickle_datasaver(x):
+    return dict(x=x, y=x)
+
+
+@flaky.flaky(max_runs=3)
+@pytest.mark.parametrize(
+    "learner_type, learner_kwargs, serializer", learners,
+)
+def test_serialization_for(learner_type, learner_kwargs, serializer):
+    """Test serializing a learner using different serializers."""
+
+    def f(x):
+        return random.random()
+
+    if serializer is pickle:
+        # f from the local scope cannot be pickled
+        f = f_for_pickle  # noqa: F811
+
+    learner = learner_type(f, **learner_kwargs)
+    if learner_type is Learner1D:
+        learner._recompute_losses_factor = 1
```

Comment:
If I understand correctly, here we're saying "we're only computing stuff approximately, so let's change that to make pickling work". I don't think that's the right way to go.

I think pickling is supposed to also store the internal state, so that we shouldn't need to tweak the learner behavior like that.

Reply:
I don't think it is necessary to require that an unpickled learner is identical to the one before. In this case, the loaded learner has a better approximation of the loss because it is able to load all points at once.

Storing all tiny details will probably mean that with the tiniest of changes (to private API), one is not able to load saved learners from a previous version.

Reply:
> In this case, the loaded learner has a better approximation of the loss because it is able to load all points at once.

That is a possible reason to deviate from the standard behavior, however, I'd also like to hear @jbweston's opinion—I still think that this is dangerously unconventional.

> Storing all tiny details will probably mean that with the tiniest of changes (even to private API), one is not able to load saved learners from a previous version.

I don't think cross-version pickling is guaranteed by anything.

Reply:
Even today I loaded data saved with Adaptive more than three years ago and it loaded without any complaint, which made me very happy. 

One other thing to keep in mind is that, even though this might not be the final answer, it's still better than the current broken state! Right now picking works for some Learners while others have a broken state after loading.

Reply:
That's the difference between data formats and pickling (preserving live objects).

One should be stable, the other should be exact.

Reply:
It is generally true that people expect `pickle.loads(pickle.dumps(x))` to be as close to the identity as possible. If we start deviating from this it is IMO almost certain that we (or one of our users) will get surprised by this at some point in the future.

I would prefer that if what someone wants to do is "create a learner with the same data" then they should export the data from the learner, pickle that, and then load it back in later.

Reply:
I am going to counter the "but this is functionally the same thing" response with "no". 

There are certain semantics associated with pickling/unpickling that people expect.

We do not set up the expectation that exporting/importing data from a learner has any given semantics beyond the most trivial (e.g. that after the import the learner has all the data specified).

Reply:
OK fair enough, I've made the appropriate changes such that the data structures are the same!

I've run the test 10_000 times and it always passes now! 🎉